Feat/new engine control

derpop opened the pull request:
## Please make sure you have done the following things.

1. Ensure that your code compiles for the correct platform. (Arduino or Linux)
2. Thoroughly tested your code to make sure it works
3. Everything that you have written is properly committed and pushed. 

## Please describe the goal of your pull request below

New Engine Control Script(UNTESTED)

derpop committed:
New rocketControl file

Remove old engineControl files, and replace with new skeleton code

derpop committed:
Updating the Engine Control 3

Its updated lmao with functions

derpop committed:
Code Cleaning

derpop committed:
Updated Readme\ Basic Logic Changes

What it says on the Tin

CooperW824 commented on README.md:
```diff
@@ -127,71 +131,61 @@ The control room and the test stand use 64 bit (8 byte) signals to communicate d
+        - Sent after the Test Stand has received the Control Room's 'PressFuel' signal, and indicates that the fuel has been pressurized.
+    - InertFlush
+        - Sent after the Test Stand has received the Control Room's 'InertFlush' signal, and indicates that that a brief burst of nitrogen was pushed through the chamber.
+    - Ignite
```
Can we make the Ignition and Ping Commands 8 bytes just to make reading acknowledgements a bit easier on the control side?

CooperW824 replied:
Same with InertFlush and PressFuel

CooperW824 replied:
We should make the Ping command's command code `PingPong` :)

CooperW824 commented on src/engineControl3/engineControl3.ino:
```diff
@@ -0,0 +1,246 @@
+    sensorData.pressure[4] = analogRead(PRESSSEN5);
+    sensorData.pressure[5] = analogRead(PRESSSEN6);
+
+    sensorData.thermocouple[0] = digitalRead(THERMOCPL1CS);
```
Thermocouples are going to be read on the analog pins, we changed the board to be an AD8495, analog amplifier. 

The analog read returns an int from 0 to 1023, maybe we should just send this data back to the control computer and let the control computer convert to Celcius:

But the overall logic is below:

const float voltageToTempScale = 0.005;  // AD8495 outputs 5mV per °C
int analogValue = analogRead(thermocouplePin);
float voltage = analogValue * (5.0 / 1023.0);  // Assuming 5V reference
float temperatureC = voltage / voltageToTempScale;

CooperW824 commented on src/engineControl3/engineControl3.ino:
```diff
@@ -0,0 +1,246 @@
+            sendSensor(sensorData);
+        }
+        if (Serial.available() >= CONTROL_ROOM_RESPONSE_SIZE) {
+            Serial.readBytes(dataIn, CONTROL_ROOM_RESPONSE_SIZE);
```
If the control room response size is 6, and the longest command is 9 bytes, you will never read the full command, so you'll never reach inert flush or fuel pressurization.

CooperW824 commented on src/engineControl3/engineControl3.ino:
```diff
@@ -0,0 +1,246 @@
+            Serial.readBytes(dataIn, CONTROL_ROOM_RESPONSE_SIZE);
+            incomingSignal = String(dataIn);
+        }
+        if(!fuelPressed && !inertFlushComplete){
```
This is going to write log start down the serial port connection forever down the line until inert flush is called, which is really not what we want because then we just have noise on the line which we then have to filter. Ideally the control computer sends the `LogStart` command, and then the test stand sends it back once. Then if we don't see the acknowledgement we should send the command again and the rocket should do nothing and resend the acknowledgement. We might need to rework this logic.

derpop replied:
This would also be the case for all other communications, i can change how they are all sent.

CooperW824 commented on src/engineControl3/engineControl3.ino:
```diff
@@ -0,0 +1,246 @@
+void loop() {
+    masterTime = millis();
+    nominalCheck();
+    sendSensor(sensorData);
```
nominal check should probably be done after `sendSensorData` to ensure the most up to date info

CooperW824 commented on README.md:
```diff
@@ -127,71 +131,61 @@ The control room and the test stand use 64 bit (8 byte) signals to communicate d
 {
   float loadCell;
-  float thermocouple;
+  float thermocouple[4];
```
we might change this to be 4 int_16 values because we may just send the raw value back to the control computer

CooperW824 replied:
this was a dumb idea of mine, we need the float values to do the nominal check so we might as well send the float value back

CooperW824 commented on README.md:
```diff
@@ -127,71 +131,61 @@ The control room and the test stand use 64 bit (8 byte) signals to communicate d
   float loadCell;
-  float thermocouple;
+  float thermocouple[4];
   float pressure[6];
```
same here this might be int_16 because its also an analog signal

CooperW824 commented on README.md:
```diff
@@ -85,7 +85,11 @@ This is built on the Arduino platform for easy and plentiful I/O. Arduino provid
 - 1 Load Cell for measuring thrust
-- 1 Thermocouple placed before the injector plate
+- 4 Thermocouples
+    - 1 Inside the Combustion Chamber
```
1. At the Nozzle exit
1. On the Kerosene Inlet and Injector Plate
1. On the N2O inlet and injector Plate
1. On the combustion chamber at the throat of the nozzle (the constriction)
![image](https://github.com/user-attachments/assets/65f1e728-bc12-4ae2-830a-f3080d1fb4bf)

CooperW824 reviewed (changes requested):
Looks good so far, just some small changes, but we may need to adjust the control flow of the engine. We can talk at the next meeting about this. But great job so far

derpop commented:
Alright I will commit those new changes by around friday at the latest.

derpop committed:
Changes Requested

Alright All changes are requested, Press Fuel and Nominal Check still not written because i dont have the data needed for these two functions.

CooperW824 reviewed (approved):
LGTM

CooperW824 committed:
fix(firmware): minor fixes for thermocouples

CooperW824 committed:
updating docs and adding method to convert analog load cell data to digital

CooperW824 committed:
Merge branch 'main' of https://github.com/Rensselaer-Spaceflight-Society/Rocket-Engine into Feat/newEngineControl
